data/transactions/logic/eval.go

```diff
@@ -472,6 +472,15 @@ func (ep *EvalParams) RecordAD(gi int, ad transactions.ApplyData) {
 	}
 }
 
+type frame struct {
+	retpc  int
+	height int
+
+	clear   bool // perform "shift and clear" in retsub
+	args    int
+	returns int
```

Comment:
I haven't looked at all the usages, but is there a reason the int's are signed?

Reply:
In Go, len() returns an `int`, not a `uint`.  I once tried to understand why, but I don't recall a much better justification than: that seems nice.

That means that whenever you have variables that are derived from, or interact with, len() values a lot, you'll be casting around like crazy if you don't go with the flow.  Since these variables interact with the stack len a lot, I think it's cleaner not to cast as much.  Here is the diff that would allow them to be uint, trimmed to just the changes.

What do you think?  It's not horrendous, but it's certainly not _more_concise. It's all susceptible to the same bugs (if you imagine, for example, that there is wraparound somewhere, it would also wrap the `uint`).

```
 type frame struct {
 	retpc  int
-	height int
+	height uint
 
 	clear   bool // perform "shift and clear" in retsub
-	args    int
-	returns int
+	args    uint
+	returns uint
 }
 
-		height: len(cx.stack),
+		height: uint(len(cx.stack)),

-		if len(cx.stack) < expect {
-			found := len(cx.stack) - frame.height
+		if uint(len(cx.stack)) < expect {
+			found := uint(len(cx.stack)) - frame.height


-	nargs := int(cx.program[cx.pc+1])
-	if nargs > len(cx.stack) {
+	nargs := uint(cx.program[cx.pc+1])
+	if nargs > uint(len(cx.stack)) {

-	cx.callstack[top].returns = int(cx.program[cx.pc+2])
+	cx.callstack[top].returns = uint(cx.program[cx.pc+2])

-	if i < 0 && -int(i) > frame.args {
+	if i < 0 && uint(-i) > frame.args {

-	idx := frame.height + int(i)
-	if idx >= len(cx.stack) {
+	idx := frame.height + uint(i)
+	if idx >= uint(len(cx.stack)) {
 
-	if i < 0 && -int(i) > frame.args {
+	if i < 0 && uint(-i) > frame.args {

-	idx := frame.height + int(i)
-	if idx >= last {
+	idx := frame.height + uint(i)
+	if idx >= uint(last) {
```

Reply:
Tough choice. Beauty or Correctness? (not that the beautiful is actually wrong in this case...) Some would say that correctness is inherently beautiful.

Usually I'd opt for beauty when there isn't really much correctness to be gained.

But how much uglier is the new code? I only count 4 new casts where there weren't any previously.

.....🥁...🥁...🥁...🥁...🥁... so I end up voting for the new version.

Reply:
I'd agree if there was any real correctness. Overflows and underflows works occur just the same. The casts feel like wishful thinking to me. Anyone else? I don't feel very strongly.

Reply:
I'm a little extra reluctant to make these `uint` while all of these are `int` in EvalContext.
```
	pc      int
	nextpc  int
	cost    int                            // cost incurred so far
	logSize int                            // total log size so far
	branchTargets map[int]bool
	instructionStarts map[int]bool
```
All of those ints are program counters or similarly non-negative things.

Reply:
Maybe that's not such a big deal - I wrote that while thinking the values in question were PCs. But they aren't, they are stack heights. I don't know, feels weird to go against the Go grain.

I closest thing to a _good_ reason is this.  Suppose you have a stack height `h`. You are contemplating subtracting x from it.  It feels very intuitive to write:
```
if h  - x < 0 {
 yell and scream
}
```
This only works if they are `int`. If `uint`, you must:
```
if h - x > h {
  yell and scream
}
```
_every single time I write that_ (which I do, when dealing with uint64 overflow) I hate it, and about 50% of the time, I get questions about it in code review.

Now, if we were counting on Go to scream at us when we _used_ that new value `h-x` we'd probably get an out of bounds exception either way. But we often want to carefully check and report a clean error.